cmd/xl-storage-disk-id-check.go

```diff
@@ -821,6 +829,83 @@ func (p *xlStorageDiskIDCheck) monitorDiskStatus() {
 	}
 }
 
+// monitorDiskStatus should be called once when a drive has been marked offline.
+// Once the disk has been deemed ok, it will return to online status.
+func (p *xlStorageDiskIDCheck) monitorDiskWritable(ctx context.Context) {
+	const (
+		// We check every 15 seconds if the disk is writable and we can read back.
+		checkEvery = 15 * time.Second
+
+		// Disk has 2 minutes to complete write+read.
+		timeoutOperation = 2 * time.Minute
+
+		// If the disk has completed an operation successfully within last 5 seconds, don't check it.
+		skipIfSuccessBefore = 5 * time.Second
+	)
+
+	t := time.NewTicker(checkEvery)
+	defer t.Stop()
+	fn := mustGetUUID()
+
+	// Be just above directio size.
+	toWrite := []byte{xioutil.DirectioAlignSize + 1: 42}
+	rng := rand.New(rand.NewSource(time.Now().UnixNano()))
+
+	for range t.C {
+		if contextCanceled(ctx) {
+			break
+		}
+		if atomic.LoadInt32(&p.health.status) != diskHealthOK {
+			continue
+		}
+		if time.Since(time.Unix(0, atomic.LoadInt64(&p.health.lastSuccess))) < skipIfSuccessBefore {
+			// We recently saw a success - no need to check.
+			continue
+		}
+		goOffline := func(err error) {
+			if atomic.CompareAndSwapInt32(&p.health.status, diskHealthOK, diskHealthFaulty) {
+				logger.LogAlwaysIf(ctx, fmt.Errorf("node(%s): taking drive %s offline: %v", globalLocalNodeName, p.storage.String(), err))
+				go p.monitorDiskStatus()
+			}
+		}
+		// Offset checks a bit.
+		time.Sleep(time.Duration(rng.Int63n(int64(1 * time.Second))))
+		done := make(chan struct{})
+		started := time.Now()
+		go func() {
+			timeout := time.NewTimer(timeoutOperation)
+			select {
+			// Reuse the same trigger
+			// If it triggers again, it took too long.
+			case <-timeout.C:
+				spent := time.Since(started)
+				goOffline(fmt.Errorf("unable to write+read for %v", spent.Round(time.Millisecond)))
+			case <-done:
+				if !timeout.Stop() {
+					<-timeout.C
+				}
+			}
+		}()
+		func() {
+			defer close(done)
+			err := p.storage.WriteAll(ctx, minioMetaTmpBucket, fn, toWrite)
+			if err != nil {
+				if osErrToFileErr(err) == errFaultyDisk {
+					goOffline(fmt.Errorf("unable to write: %w", err))
+				}
+				return
+			}
+			b, err := p.storage.ReadAll(context.Background(), minioMetaTmpBucket, fn)
```

Comment:
We should also add Delete() perhaps like how we do the regular monitoring?

Reply:
I didn't really see a need, since it will be potentially leaky on restarts anyway. Trying to do *min*IO here :D

I do have a Q, though - we are clearing `.minio.sys/tmp` on every restart and not just `.minio.sys/tmp/.trash`, right?

I didn't want to place it in `.trash`, since we would be racing the cleanup routine - but I wanted it to be deleted on restarts, since it will be leaky (and we can't use a deterministic name, since at least two instances will be doing the operations).

Reply:
Yes correct upon every restart.

Reply:
Yeah. No need to delete then.